File: data_structures/hash_table/typescript/HashTable.ts
import DoublyLinkedList from '@ds/linked_list/typescript/DoublyLinkedList';

const encoding = new TextEncoder();
export type KeyType = string | number;

/**
 * Table
 * {
 *   'a' => { value: 97: order: 0 }
 * }
 *
 * KeysArray = [
 *    0 = 'a'
 * ]
 *
 * KeysIndex = 0
 *
 */

type HashTableRecord<T> = { value: T; order: number };

export default class HashTable<T, K = unknown> {
    size = 0;
    loadFactor: number;
    table: (DoublyLinkedList<HashTableRecord<K>> | null)[];
    initialCapacity: number;
    uniqueIdPropName = '425123g2e';
    objectIdCounter = -1;
    keysArray: T[] = [];
    keysIndex = 0;

    constructor(initialCapacity = 57, loadFactor = 0.75) {
        this.loadFactor = loadFactor;
        this.initialCapacity = initialCapacity;
        this.reset();
    }

    reset(buckets = new Array(this.initialCapacity), size = 0, keysArray = [], keysIndex = 0): void {
        this.table = buckets;
        this.size = size;
        this.keysArray = keysArray;
        this.keysIndex = keysIndex;
    }

    put(key: T, value: K): this {
        const hash = this.hash(key);
        const bucket = this.table[hash];

        if (!bucket) {
            this.table[hash] = new DoublyLinkedList<HashTableRecord<K>>().push({ value, order: this.keysIndex }, hash);
            this.keysArray[this.keysIndex] = key;
            this.keysIndex += 1;
            this.size++;
        } else {
            let current = bucket.head;
            while (current) {
                if (current.key == hash) {
                    current.val.value = value; // update
                    return this;
                } else {
                    current = current.next;
                }
            }
            // accommodate collision
            bucket.push({ value, order: this.keysIndex }, hash);
            this.size++;
        }

        return this;
    }

    get(key: T): null | K {
        const hash = this.hash(key);
        const bucket = this.table[hash];

        if (!bucket) {
            return null;
        }

        const node = bucket.find(hash);
        if (!node) {
            return null;
        }

        return node.val.value;
    }

    has(key: T): boolean {
        const node = this.get(key);
        return !!node;
    }

    remove(key: T): boolean {
        const hash = this.hash(key);
        const bucket = this.table[hash];

        if (!bucket) {
            return false;
        }

        const node = bucket.find(hash);
        if (!node) {
            return false;
        }

        bucket.delete(node);
        delete this.keysArray[node.val.order];
        this.size--;

        return true;
    }

    hash(key: T): number {
        let toHash;

        if (typeof key === 'object') {
            toHash = this.getObjectUniqueId(key);
        } else {
            toHash = key;
        }

        const bytes = encoding.encode(String(toHash));
        let hash = 0;
        for (const char of bytes) {
            hash ^= char;
            hash *= 16777619;
        }

        return (hash >>> 0) % this.initialCapacity;
    }

    fromArray(keyValues: Array<[T, K]>): this {
        if (!keyValues || keyValues.length == 0) {
            return this;
        }

        for (const keyVal of keyValues) {
            const key = keyVal[0];
            const value = keyVal[1];

            this.put(key, value);
        }

        return this;
    }

    get nextObjectId(): number {
        return this.objectIdCounter++;
    }

    getObjectUniqueId(object: unknown): number {
        if (object == null) {
            return null;
        }

        const id = object[this.uniqueIdPropName];

        if (id !== null) {
            return id;
        }

        if (Object.isFrozen(object)) {
            return null;
        }

        const uniqueId = this.nextObjectId;

        Object.defineProperty(object, this.uniqueIdPropName, {
            enumerable: false,
            configurable: false,
            writable: false,
            value: uniqueId,
        });

        return uniqueId;
    }

    rehash(): void {
        throw new Error('Not implemented');
    }

    *keys(): Iterable<T> {
        for (let i = 0; i < this.keysArray.length; i++) {
            const key = this.keysArray[i];
            if (key !== undefined) {
                yield key;
            }
        }
    }

    *values(): Iterable<K> {
        for (const key of this.keys()) {
            yield this.get(key);
        }
    }
}

File: data_structures/hash_table/python/hash_table.py
from typing import List, TypeVar, Generic, Union, Tuple, Any
from data_structures.linked_list.python.doubly_linked_list import DoublyLinkedList

T = TypeVar('T')
K = TypeVar('K')

HashTableType = List[Union[DoublyLinkedList[T], None]]


class HashTable(Generic[T]):
    table: HashTableType
    size: int

    def __init__(self, size_table: int = 64):
        self.table = [None] * size_table
        self.size = size_table

    def hash(self, key: Any) -> int:
        key = bytearray(str(key).encode())

        h_key = 0
        for char in key:
            h_key ^= char  # XOR
            h_key *= 16777619 # 32 bit prime number

        # zero-fill right shift
        h_key = (h_key >> 0) if h_key >= 0 else ((h_key + 0x100000000) >> 0)

        return h_key % self.size

    def put(self, key: Any, value: T) -> 'HashTable':
        hash_code = self.hash(key)
        bucket = self.table[hash_code]

        if bucket is None:
            self.table[hash_code] = DoublyLinkedList().push(key=key, val=value)
        else:
            current = bucket.head
            while current:
                if current.key == key:
                    current.val = value  # update
                    return self
                else:
                    current = current.next

            bucket.push(key=key, val=value)

        return self

    def get(self, key: Any) -> Union[None, T]:
        hash_code = self.hash(key)
        bucket = self.table[hash_code]

        if bucket is None:
            return None

        node = bucket.find(key)
        if node is None:
            return None

        return node.val

    def remove(self, key: Any) -> Union[None, T]:
        hash_code = self.hash(key)
        bucket = self.table[hash_code]

        if bucket is None:
            return None

        node = bucket.find(key)
        bucket.delete(node)

        return node.val

    def from_array(self, key_values: List[Tuple[Any,T]]) -> 'HashTable':
        if len(key_values) == 0:
            return self

        for t in key_values:
            key, value = t[0], t[1]
            self.put(key, value)

        return self

    def __repr__(self):
        result = ""
        for unit in self.table:
            result += "\t" + repr(unit) + "\n"
        return f"HashTable[\n{result}]"
